File: python/alphaarc/correspondence.py
"""Correspondence / Copy -- relational compression. Port of pkg/macro/correspondence.go.

Region A described as "a copy of region B up to {Identity, ColorSwap, Reflect}
plus a residual". Closes the template/legend/copy class (s5i5, tn36, vc33-L2):
puzzles that are anti-compression to self-regularity but compression-positive
relationally. Search stays O(objects): anchors = repeated components grouped by a
loose (WxH, colour) signature, only within-class members compared, constant
transform set. Importing this module REGISTERS Correspondence in mdl.PRIMITIVES.
"""

from __future__ import annotations

from typing import Dict, List, Tuple

from .mdl import Grid, Primitive, PRIMITIVES, _components

Cell = Tuple[int, int]  # (row, col)

_POINTER_COST = 1   # bits to name template + anchor (small constant)
_MIN_AREA = 4       # a real region, not a dot/line (degenerate-cheat guard)


def _cells_bbox(cells: List[Cell]) -> Tuple[int, int, int, int]:
    rs = [c[0] for c in cells]
    cs = [c[1] for c in cells]
    return min(rs), min(cs), max(rs), max(cs)


def _subgrid(grid: Grid, minr: int, minc: int, maxr: int, maxc: int) -> Grid:
    out = []
    for r in range(minr, maxr + 1):
        row = []
        for c in range(minc, maxc + 1):
            row.append(grid[r][c] if r < len(grid) and c < len(grid[r]) else 0)
        out.append(row)
    return out


def _distinct_colors(g: Grid) -> int:
    return len({v for row in g for v in row})


def _residual_identity(a: Grid, b: Grid) -> int:
    return sum(1 for i in range(len(a)) for j in range(len(a[i])) if a[i][j] != b[i][j])


def _residual_reflect_h(a: Grid, b: Grid) -> int:
    res = 0
    for i in range(len(a)):
        w = len(a[i])
        for j in range(w):
            if a[i][j] != b[i][w - 1 - j]:
                res += 1
    return res


def _residual_reflect_v(a: Grid, b: Grid) -> int:
    res = 0
    h = len(a)
    for i in range(h):
        for j in range(len(a[i])):
            if a[i][j] != b[h - 1 - i][j]:
                res += 1
    return res


def _color_swap_map(a: Grid, b: Grid, bg: int) -> Dict[int, int]:
    """Majority map b->a in O(area). bg pinned to bg (a hole is a deletion, not a recolour)."""
    tally: Dict[int, Dict[int, int]] = {}
    for i in range(len(a)):
        for j in range(len(a[i])):
            bc, ac = b[i][j], a[i][j]
            if bc == bg:
                continue
            tally.setdefault(bc, {})
            tally[bc][ac] = tally[bc].get(ac, 0) + 1
    m = {bg: bg}
    for bc, counts in tally.items():
        best, bn = bc, -1
        for ac, n in counts.items():
            if ac == bg:
                continue
            if n > bn or (n == bn and ac < best):
                best, bn = ac, n
        m[bc] = best
    return m


def _residual_color_swap(a: Grid, b: Grid, bg: int) -> int:
    m = _color_swap_map(a, b, bg)
    return sum(1 for i in range(len(a)) for j in range(len(a[i])) if a[i][j] != m.get(b[i][j], b[i][j]))


def _min_residual(a: Grid, b: Grid, bg: int) -> int:
    return min(
        _residual_identity(a, b),
        _residual_color_swap(a, b, bg),
        _residual_reflect_h(a, b),
        _residual_reflect_v(a, b),
    )


# --- *_cells variants: the LOCAL positions in `a` that differ (click candidates) ---

def _residual_identity_cells(a: Grid, b: Grid) -> List[Cell]:
    return [(i, j) for i in range(len(a)) for j in range(len(a[i])) if a[i][j] != b[i][j]]


def _residual_reflect_h_cells(a: Grid, b: Grid) -> List[Cell]:
    out = []
    for i in range(len(a)):
        w = len(a[i])
        for j in range(w):
            if a[i][j] != b[i][w - 1 - j]:
                out.append((i, j))
    return out


def _residual_reflect_v_cells(a: Grid, b: Grid) -> List[Cell]:
    out = []
    h = len(a)
    for i in range(h):
        for j in range(len(a[i])):
            if a[i][j] != b[h - 1 - i][j]:
                out.append((i, j))
    return out


def _residual_color_swap_cells(a: Grid, b: Grid, bg: int) -> List[Cell]:
    m = _color_swap_map(a, b, bg)
    return [(i, j) for i in range(len(a)) for j in range(len(a[i])) if a[i][j] != m.get(b[i][j], b[i][j])]


def _min_residual_cells(a: Grid, b: Grid, bg: int) -> List[Cell]:
    best = _residual_identity_cells(a, b)
    for cells in (
        _residual_color_swap_cells(a, b, bg),
        _residual_reflect_h_cells(a, b),
        _residual_reflect_v_cells(a, b),
    ):
        if len(cells) < len(best):
            best = cells
    return best


def _groups(grid: Grid, bg: int) -> Dict[str, List[Tuple[int, int, int, int]]]:
    """Anchors: component bboxes grouped by the loose (WxH, colour) signature."""
    comps = _components(grid, bg)
    if len(comps) < 2:
        return {}
    groups: Dict[str, List[Tuple[int, int, int, int]]] = {}
    for color, cells in comps:
        minr, minc, maxr, maxc = _cells_bbox(cells)
        w, h = maxc - minc + 1, maxr - minr + 1
        if w * h < _MIN_AREA:
            continue
        key = "%dx%d:c%d" % (w, h, color)
        groups.setdefault(key, []).append((minr, minc, maxr, maxc))
    return groups


def _reference(groups_members):
    """Extract member subgrids and pick the fullest (most non-bg) as the reference."""
    return groups_members


def correspondence_savings(grid: Grid, bg: int) -> int:
    savings = 0
    for members in _groups(grid, bg).values():
        if len(members) < 2:
            continue
        grids = [_subgrid(grid, *m) for m in members]
        ref_idx, ref_full = 0, -1
        for i, g in enumerate(grids):
            full = sum(1 for row in g for v in row if v != bg)
            if full > ref_full:
                ref_full, ref_idx = full, i
        ref = grids[ref_idx]
        if _distinct_colors(ref) < 2:
            continue  # reference must be a real pattern, not a solid block
        area = len(ref) * len(ref[0])
        for i, g in enumerate(grids):
            if i == ref_idx:
                continue
            s = area - _min_residual(g, ref, bg) - _POINTER_COST
            if s > 0:
                savings += s
    return savings


def correspondence_preference(grid: Grid, bg: int) -> int:
    return correspondence_savings(grid, bg)


def correspondence_residual(grid: Grid, bg: int) -> List[Cell]:
    """ABSOLUTE grid cells that disagree with their class reference -- click candidates."""
    out: List[Cell] = []
    for members in _groups(grid, bg).values():
        if len(members) < 2:
            continue
        grids = [_subgrid(grid, *m) for m in members]
        ref_idx, ref_full = 0, -1
        for i, g in enumerate(grids):
            full = sum(1 for row in g for v in row if v != bg)
            if full > ref_full:
                ref_full, ref_idx = full, i
        ref = grids[ref_idx]
        if _distinct_colors(ref) < 2:
            continue
        for i, g in enumerate(grids):
            if i == ref_idx:
                continue
            minr, minc = members[i][0], members[i][1]
            for (cr, cc) in _min_residual_cells(g, ref, bg):
                out.append((minr + cr, minc + cc))
    return out


# Register Correspondence as the 4th primitive (LAST, matching Go's order so
# BestPrimitive ties resolve identically). Idempotent under re-import.
if not any(p.name == "Correspondence" for p in PRIMITIVES):
    PRIMITIVES.append(Primitive("Correspondence", correspondence_preference))

File: pkg/macro/correspondence.go
package macro

import "fmt"

// Primitive #4: Correspondence / Copy (RELATIONAL compression).
//
// Reflect/Translate/Count measure SELF-regularity (a region regular within
// itself). Correspondence measures regularity BETWEEN regions: region A described
// as "a copy of region B, up to a simple transform, plus a residual". This closes
// the copy/match/template class (s5i5's template boxes, tn36's legend, vc33-L2's
// mirror flip) -- puzzles whose solution is ANTI-compression to a self-regularity
// drive (filling a template raises A's own complexity) but compression-POSITIVE
// relationally (A becomes describable as B, so the residual shrinks).
//
// Search stays O(objects), no NP explosion:
//   - ANCHORS are repetitions, found by the SAME component grouping as Count but
//     with a LOOSE signature (bbox WxH + border/component colour) so regions with
//     different interiors still cluster (V1: exact WxH, no scale).
//   - only WITHIN-class members are compared (k small), never all-pairs.
//   - a CONSTANT transform set {Identity, ColorSwap, Reflect(H/V)} -- ColorSwap's
//     mapping is derived by an O(area) majority vote, not searched.
//
// Saving(class) = sum over non-reference members A of
//     area - residual(A, ref) - pointerCost
// so filling one correct cell of a partial pattern drops the residual by 1 and
// lifts the saving by 1 -- the dopamine click on DriveScore.

const correspondencePointerCost = 1 // bits to name the template + anchor (small constant)
const correspondenceMinArea = 4     // a real region, not a dot/line (degenerate-cheat guard)

func cellsBBox(cells []Cell) (minR, minC, maxR, maxC int) {
	minR, minC = cells[0].R, cells[0].C
	maxR, maxC = cells[0].R, cells[0].C
	for _, cl := range cells {
		if cl.R < minR {
			minR = cl.R
		}
		if cl.R > maxR {
			maxR = cl.R
		}
		if cl.C < minC {
			minC = cl.C
		}
		if cl.C > maxC {
			maxC = cl.C
		}
	}
	return
}

// subgridOf extracts the full rectangular content (all colours) of a bbox.
func subgridOf(grid [][]int, minR, minC, maxR, maxC int) [][]int {
	out := make([][]int, maxR-minR+1)
	for r := minR; r <= maxR; r++ {
		row := make([]int, maxC-minC+1)
		for c := minC; c <= maxC; c++ {
			if r < len(grid) && c < len(grid[r]) {
				row[c-minC] = grid[r][c]
			}
		}
		out[r-minR] = row
	}
	return out
}

func distinctColorCount(g [][]int) int {
	seen := map[int]struct{}{}
	for _, row := range g {
		for _, v := range row {
			seen[v] = struct{}{}
		}
	}
	return len(seen)
}

func residualIdentity(a, b [][]int) int {
	res := 0
	for i := range a {
		for j := range a[i] {
			if a[i][j] != b[i][j] {
				res++
			}
		}
	}
	return res
}

func residualReflectH(a, b [][]int) int {
	res := 0
	for i := range a {
		w := len(a[i])
		for j := 0; j < w; j++ {
			if a[i][j] != b[i][w-1-j] {
				res++
			}
		}
	}
	return res
}

func residualReflectV(a, b [][]int) int {
	res := 0
	h := len(a)
	for i := 0; i < h; i++ {
		for j := range a[i] {
			if a[i][j] != b[h-1-i][j] {
				res++
			}
		}
	}
	return res
}

// residualColorSwap derives the majority colour map b->a in O(area) (no search)
// and counts the residual under it. Subsumes Identity when the map is the
// identity, and catches a pure recolouring of the same pattern. bg is pinned to
// bg -- a background hole is a deletion, NOT a recolouring, so ColorSwap may not
// remap bg to a foreground colour (which would hide a missing cell).
// colorSwapMap derives the majority colour map b->a in O(area) (no search).
// bg is pinned to bg -- a background hole is a deletion, NOT a recolouring,
// so ColorSwap may not remap bg to a foreground colour (which would hide a
// missing cell), nor remap a foreground colour onto bg.
func colorSwapMap(a, b [][]int, bg int) map[int]int {
	tally := map[int]map[int]int{}
	for i := range a {
		for j := range a[i] {
			bc, ac := b[i][j], a[i][j]
			if bc == bg {
				continue
			}
			if tally[bc] == nil {
				tally[bc] = map[int]int{}
			}
			tally[bc][ac]++
		}
	}
	m := map[int]int{bg: bg}
	for bc, counts := range tally {
		best, bn := bc, -1
		for ac, n := range counts {
			if ac == bg {
				continue
			}
			if n > bn || (n == bn && ac < best) {
				best, bn = ac, n
			}
		}
		m[bc] = best
	}
	return m
}

// residualColorSwap counts the residual under colorSwapMap's derived mapping.
// Subsumes Identity when the map is the identity, and catches a pure
// recolouring of the same pattern.
func residualColorSwap(a, b [][]int, bg int) int {
	m := colorSwapMap(a, b, bg)
	res := 0
	for i := range a {
		for j := range a[i] {
			if a[i][j] != m[b[i][j]] {
				res++
			}
		}
	}
	return res
}

// minResidual is the fewest cells A needs to differ from B under any allowed
// transform {Identity, ColorSwap, Reflect-H, Reflect-V}.
func minResidual(a, b [][]int, bg int) int {
	best := residualIdentity(a, b)
	for _, r := range []int{residualColorSwap(a, b, bg), residualReflectH(a, b), residualReflectV(a, b)} {
		if r < best {
			best = r
		}
	}
	return best
}

// The *Cells variants below mirror residualIdentity/residualReflectH/
// residualReflectV/residualColorSwap exactly, but return the LOCAL (row,col)
// positions inside `a` that differ, instead of just a count -- the click
// candidates for correspondenceResidual. Kept separate from the hot-path
// count-only functions (called once per grid, every candidate primitive
// comparison) to avoid allocating cell slices there.

func residualIdentityCells(a, b [][]int) []Cell {
	var out []Cell
	for i := range a {
		for j := range a[i] {
			if a[i][j] != b[i][j] {
				out = append(out, Cell{R: i, C: j})
			}
		}
	}
	return out
}

func residualReflectHCells(a, b [][]int) []Cell {
	var out []Cell
	for i := range a {
		w := len(a[i])
		for j := 0; j < w; j++ {
			if a[i][j] != b[i][w-1-j] {
				out = append(out, Cell{R: i, C: j})
			}
		}
	}
	return out
}

func residualReflectVCells(a, b [][]int) []Cell {
	var out []Cell
	h := len(a)
	for i := 0; i < h; i++ {
		for j := range a[i] {
			if a[i][j] != b[h-1-i][j] {
				out = append(out, Cell{R: i, C: j})
			}
		}
	}
	return out
}

func residualColorSwapCells(a, b [][]int, bg int) []Cell {
	m := colorSwapMap(a, b, bg)
	var out []Cell
	for i := range a {
		for j := range a[i] {
			if a[i][j] != m[b[i][j]] {
				out = append(out, Cell{R: i, C: j})
			}
		}
	}
	return out
}

// minResidualCells returns the local (row,col) positions inside `a` that
// differ from `b` under whichever transform gives the FEWEST differences --
// matching minResidual's choice exactly, so this is "the cells that need to
// change for `a` to become a copy of `b`" under the same transform the
// savings computation already picked.
func minResidualCells(a, b [][]int, bg int) []Cell {
	best := residualIdentityCells(a, b)
	for _, cells := range [][]Cell{residualColorSwapCells(a, b, bg), residualReflectHCells(a, b), residualReflectVCells(a, b)} {
		if len(cells) < len(best) {
			best = cells
		}
	}
	return best
}

// nonBgCount counts non-background cells in a region.
func nonBgCount(g [][]int, bg int) int {
	n := 0
	for _, row := range g {
		for _, v := range row {
			if v != bg {
				n++
			}
		}
	}
	return n
}

// correspondenceRegion is one anchor member's bbox in absolute grid coords.
type correspondenceRegion struct{ minR, minC, maxR, maxC int }

// correspondenceGroups finds ANCHORS: components grouped by the loose
// (W, H, colour) signature (V1: exact WxH, no scale) -- the same grouping
// CorrespondenceSavings and correspondenceResidual both need.
func correspondenceGroups(grid [][]int, bg int) map[string][]correspondenceRegion {
	comps := components(grid, bg)
	if len(comps) < 2 {
		return nil
	}
	groups := map[string][]correspondenceRegion{}
	for _, cp := range comps {
		minR, minC, maxR, maxC := cellsBBox(cp.cells)
		w, h := maxC-minC+1, maxR-minR+1
		if w*h < correspondenceMinArea {
			continue
		}
		key := fmt.Sprintf("%dx%d:c%d", w, h, cp.color)
		groups[key] = append(groups[key], correspondenceRegion{minR, minC, maxR, maxC})
	}
	return groups
}

// CorrespondenceSavings sums the description saved by describing repeated
// same-size regions as copies of one representative (up to a simple transform).
func CorrespondenceSavings(grid [][]int, bg int) int {
	groups := correspondenceGroups(grid, bg)
	savings := 0
	for _, members := range groups {
		if len(members) < 2 {
			continue
		}
		// Extract each member's full bbox content; the reference is the FULLEST
		// member (most non-bg cells) so a partial copy is scored against the
		// template, not vice-versa (a bg hole must not become the reference and get
		// remapped away).
		grids := make([][][]int, len(members))
		refIdx, refFull := 0, -1
		for i, m := range members {
			grids[i] = subgridOf(grid, m.minR, m.minC, m.maxR, m.maxC)
			if full := nonBgCount(grids[i], bg); full > refFull {
				refFull, refIdx = full, i
			}
		}
		ref := grids[refIdx]
		if distinctColorCount(ref) < 2 {
			continue // reference must be a real pattern, not a solid block (cheat guard)
		}
		area := len(ref) * len(ref[0])
		for i, g := range grids {
			if i == refIdx {
				continue
			}
			s := area - minResidual(g, ref, bg) - correspondencePointerCost
			if s > 0 {
				savings += s
			}
		}
	}
	return savings
}

// correspondenceResidual returns the ABSOLUTE grid cells that currently
// disagree with their class's reference (template) under the best-fitting
// transform -- the cells a click would need to fix to raise
// CorrespondenceSavings. This is the click-candidate feed for ResidualCells:
// Reflect/Translate/Count's residual is a DIFFERENT notion of anomaly
// (self-regularity broken), which routinely does NOT include the cell that
// matters for a relational/template puzzle (tn36's legend, s5i5's boxes, and
// the same pattern found live on ft09-0d8bbf25: Correspondence savings=89,
// nonzero, but zero candidate clicks ever targeted it in 150 real actions,
// because ResidualCells never had a case for Correspondence at all).
func correspondenceResidual(grid [][]int, bg int) []Cell {
	groups := correspondenceGroups(grid, bg)
	var out []Cell
	for _, members := range groups {
		if len(members) < 2 {
			continue
		}
		grids := make([][][]int, len(members))
		refIdx, refFull := 0, -1
		for i, m := range members {
			grids[i] = subgridOf(grid, m.minR, m.minC, m.maxR, m.maxC)
			if full := nonBgCount(grids[i], bg); full > refFull {
				refFull, refIdx = full, i
			}
		}
		ref := grids[refIdx]
		if distinctColorCount(ref) < 2 {
			continue
		}
		for i, g := range grids {
			if i == refIdx {
				continue
			}
			for _, c := range minResidualCells(g, ref, bg) {
				out = append(out, Cell{R: members[i].minR + c.R, C: members[i].minC + c.C})
			}
		}
	}
	return out
}

// CorrespondencePreference is the compression saving under the Copy primitive.
func CorrespondencePreference(grid [][]int, bg int) int {
	return CorrespondenceSavings(grid, bg)
}
